### PR TITLE
Feat virtio rng rate limiter

### PR DESCRIPTION
**Pull request by bchalios**
## Changes

Add rate limiting for the virtio-rng device. The properties of the rate-limiter are defined in the `/entropy` API call through a [RateLimiter](https://github.com/firecracker-microvm/firecracker/blob/34570dc7912de2d49a99e2f995e763a90236202c/src/api_server/swagger/firecracker.yaml#L1081) optional object.

If this is not defined, then the device will be operated without a rate-limiter.

## Reason

Requests for entropy from the guest consume host resource (currently we perform a `getrandom` system call). In production environments users might want to restrict the amount of entropy provided to each guest.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [ ] If a specific issue led to this PR, this PR closes the issue.
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] API changes follow the [Runbook for Firecracker API changes][2].
- [ ] User-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.
- [ ] New `TODO`s link to an issue.
- [ ] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [ ] This functionality can be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md

### CODE REVIEW

**Comment by bchalios** [comment]
This is currently rebased on top of #3254, which should be merged first!

**Comment by bchalios** [comment]
> This is currently rebased on top of #3254, which should be merged first!

#3254 has now been merged!

**Comment by bchalios** [comment]
API changes documentation and README entries will be provided with a follow-up PR when all virtio-rng related changes have been merged on the feature branch

File: src/devices/src/virtio/rng/device.rs [hunk 1]
```diff
@@ -237,7 +292,10 @@ impl VirtioDevice for Entropy {
 
 #[cfg(test)]
 mod tests {
+    use std::time::Duration;
+
     use super::*;
+    use crate::check_metric_after_block;
```
Comment (roypat) [inline src/devices/src/virtio/rng/device.rs:298]: everytime I see this function I have to do a double take because I think its related to the block device 😂

File: src/devices/src/virtio/rng/device.rs [hunk 2]
```diff
@@ -404,9 +483,135 @@ mod tests {
...
+            METRICS.entropy.entropy_event_count.count(),
+            entropy_event_count + 1
+        );
+        assert_eq!(METRICS.entropy.entropy_bytes.count(), entropy_bytes + 512);
+        assert_eq!(METRICS.entropy.host_rng_fails.count(), host_rng_fails);
+    }
+
+    #[test]
+    fn test_bad_rate_limiter_event() {
+        let mem = create_virtio_mem();
```
Comment (roypat) [inline src/devices/src/virtio/rng/device.rs:505]: Nit: Could we use `single_region_mem` from the same `test_utils` module instead here?

File: src/devices/src/virtio/rng/device.rs [hunk 3]
```diff
@@ -404,9 +483,135 @@ mod tests {
...
+        th.add_desc_chain(RNG_QUEUE, 0, &[(0, 4000, VIRTQ_DESC_F_WRITE)]);
+        check_metric_after_block!(
+            METRICS.entropy.entropy_bytes,
+            4000,
+            th.device().process_entropy_queue()
+        );
+        assert!(!th.device().rate_limiter.is_blocked());
+
+        // Sleep for 1 second to completely replenish the rate limiter
+        std::thread::sleep(Duration::from_millis(1000));
```
Comment (roypat) [inline src/devices/src/virtio/rng/device.rs:538]: Nit: Could we just call `RateLimiter::manual_replenish` here? Just in case there's some weird timing conditions that could cause spurious failures.

File: src/api_server/swagger/firecracker.yaml [hunk 4]
```diff
@@ -464,6 +464,13 @@ paths:
       description:
         Enables an entropy device that provides high-quality random data to the guest.
       operationId: putEntropyDevice
+      parameters:
+        - name: body
+          in: body
+          description: Guest entropy device properties
+          required: false
```
Comment (roypat) [inline src/api_server/swagger/firecracker.yaml:471]: I don't really know my way around the API server, but is the body of this PUT request really optional? This code 

```rs
pub(crate) fn parse_put_entropy(body: &Body) -> Result<ParsedRequest, Error> {
    let cfg = serde_json::from_slice::<EntropyDeviceConfig>(body.raw())?;
    Ok(ParsedRequest::new_sync(VmmAction::SetEntropyDevice(cfg)))
}
```

seems to imply that its not (e.g. I can't just put `null` in the body)

File: src/devices/src/virtio/rng/event_handler.rs [hunk 5]
```diff
@@ -57,6 +60,7 @@ impl MutEventSubscriber for Entropy {
 
         if !event_set.contains(EventSet::IN) {
             warn!("entropy: Received unknowon event: {event_set:?} from source {source}");
```
Comment (roypat) [inline src/devices/src/virtio/rng/event_handler.rs:62]: nit: "unknowon" -> "unknown"

File: src/devices/src/virtio/rng/device.rs (hunk 3 above)
Comment (bchalios) [reply src/devices/src/virtio/rng/device.rs:538]: I dislike it too and I think in this particular case we can do it. But these things are all over the place and I'm not sure we can get rid of them. I 'm a bit worried about having the rate limiter replenishing itself and us doing `manual_replenish` will make this logic quite hard to follow.

File: src/api_server/swagger/firecracker.yaml (hunk 4 above)
Comment (bchalios) [reply src/api_server/swagger/firecracker.yaml:471]: I think that what does the trick here is that `EntropyDeviceConfig` includes a `Option<RateLimiterConfig>`.

File: src/devices/src/virtio/rng/event_handler.rs (hunk 5 above)
Comment (bchalios) [reply src/devices/src/virtio/rng/event_handler.rs:62]: ty!

File: src/devices/src/virtio/rng/device.rs (hunk 3 above)
Comment (roypat) [reply src/devices/src/virtio/rng/device.rs]: Ah I see, yeah I didn't take that into consideration. Makes sense to keep it this way then

File: src/devices/src/virtio/rng/device.rs (hunk 2 above)
Comment (roypat) [reply src/devices/src/virtio/rng/device.rs:505]: Fair, I guess the prefered solution would then be to update all other tests to use `create_virtio_mem` and that's not in-scope here 👍

File: src/api_server/swagger/firecracker.yaml (hunk 4 above)
Comment (roypat) [reply src/api_server/swagger/firecracker.yaml:471]: Mh, I'm not sure, there's a difference between having `body = {rate_limiter: null}` and `body = null`

Comment (bchalios) [reply src/api_server/swagger/firecracker.yaml:471]: hmmm, you are right. Taking a look.

Comment (bchalios) [reply src/api_server/swagger/firecracker.yaml]: Ok fixed!

**Review by roypat — APPROVED** [review]

**Review by roypat — APPROVED** [review]

File: src/devices/src/virtio/rng/device.rs [hunk 6]
```diff
@@ -43,15 +44,18 @@ pub struct Entropy {
     queues: Vec<Queue>,
     queue_events: Vec<EventFd>,
     irq_trigger: IrqTrigger,
+
+    // Device specific fields
+    rate_limiter: RateLimiter,
```
Comment (JonathanWoollett-Light) [inline src/devices/src/virtio/rng/device.rs:49]: Nit: This could be a doc comment.
```suggestion
    /// Device specific fields
    rate_limiter: RateLimiter,
```

**Review by JonathanWoollett-Light — APPROVED** [review]

File: src/vmm/src/vmm_config/entropy.rs [hunk 7]
```diff
@@ -90,7 +107,7 @@ mod tests {
     #[test]
     fn test_set_device() {
         let mut builder = EntropyDeviceBuilder::new();
-        let device = Entropy::new().unwrap();
+        let device = Entropy::new(RateLimiter::default()).unwrap();
```
Comment (JonathanWoollett-Light) [inline src/vmm/src/vmm_config/entropy.rs:110]: Nit: Could `Entropy` implement `Default`?
```suggestion
        let device = Entropy::default().unwrap();
```
